### ISSUE TITLE
Hashable builtin and user-defined types should be subtypes of Hashable

### ISSUE THREAD

**Issue by pkch** [python/mypy#3150]
Perhaps I don't understand the purpose of `typing.Hashable`, but I expected this code to be type check:

```
K = TypeVar("K", bound=Hashable)

class MySet(AbstractSet[K]):
    def __init__(self, item: K) -> None:
        self.item = item
    def __len__(self) -> int: return 1
    def __contains__(self, item: object) -> bool: return item == self.item
    def __iter__(self) -> Iterator[K]: yield self.item

x = MySet(1) # Type argument 1 of "MySet" has incompatible value "int"

class A: ...
a = MySet(A()) # Type argument 1 of "MySet" has incompatible value "A"

class B(Hashable): ...
b = MySet(B()) # Cannot instantiate abstract class 'B' with abstract attribute '__hash__'
```

In other words, even though runtime knows that `int`, and an arbitrary used-defined class (by default) are hashable (and so they pass `issubclass(int, collections.abc.Hashable)`, etc.), mypy doesn't seem to know that. It rejects all three attempts because `int`, `A` are not subclasses of `Hashable`, and `B` because it does not have `__hash__` method explicitly defined (which I don't think is right, since `hash(B())` doesn't raise, and `issubclass(B, Hashable)` is `True`).

The case with `B` is harder, since python runtime raises too, complainign that `B` doesn't have `__hash__` method. That's really incorrect, IMO, since `B().__hash__()` does not raise; but since this issue is for `mypy`, not for core python, perhaps we shouldn't worry about it here.

But IMO, mypy should know which builtin classes are hashable. Also, when a user-defined class does not set `__hash__ = None`, mypy should also assume it's hashable.

The only way to make this work is:

```
class C(Hashable):
    def __hash__(self) -> int: return 0
c = MySet(C())
```

but that's seems unnecessary if I'm happy with the default hashing mechanism. Also, it doesn't help in the case of `int`.

A separate, but related issue is whether `bound=Hashable` is even needed. The program type checked without it, but that doesn't seem right since `AbstractSet` should really demand a hashable type as its type parameter.

**Comment by ilevkivskyi** [comment]
This will be fixed by protocols. Also this looks like a duplicate of https://github.com/python/mypy/issues/1746 and two other open issues about ``Hashable``.
